abci/app.go

```diff
@@ -0,0 +1,378 @@
+package forum
+
+import (
+	"context"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/alijnmerchant21/forum-updated/model"
+
+	"github.com/cometbft/cometbft/abci/types"
+	abci "github.com/cometbft/cometbft/abci/types"
+	cryptoencoding "github.com/cometbft/cometbft/crypto/encoding"
+	cryptoproto "github.com/cometbft/cometbft/proto/tendermint/crypto"
+
+	"github.com/cometbft/cometbft/version"
+	"github.com/dgraph-io/badger/v3"
+)
+
+const ApplicationVersion = 1
+
+type ForumApp struct {
+	abci.BaseApplication
+	valAddrToPubKeyMap map[string]cryptoproto.PublicKey
+	CurseWords         string
+	state              AppState
+	onGoingBlock       *badger.Txn
+}
+
+func NewForumApp(dbDir string, appConfigPath string) (*ForumApp, error) {
+
+	db, err := model.NewDB(dbDir)
+	if err != nil {
+		fmt.Printf("Error initializing database: %s\n", err)
+		return nil, err
+	}
+	cfg, err := LoadConfig(appConfigPath)
+	if err != nil {
+		cfg = new(Config)
+		cfg.CurseWords = "bad"
+	}
+	return &ForumApp{
+		state:              loadState(db),
+		valAddrToPubKeyMap: make(map[string]cryptoproto.PublicKey),
+		CurseWords:         cfg.CurseWords,
+	}, nil
+
+}
+
+// Return application info
+func (app ForumApp) Info(_ context.Context, info *abci.RequestInfo) (*abci.ResponseInfo, error) {
+
+	if len(app.valAddrToPubKeyMap) == 0 && app.state.Height > 0 {
+		validators := app.getValidators()
+		for _, v := range validators {
+			pubkey, err := cryptoencoding.PubKeyFromProto(v.PubKey)
+			if err != nil {
+				panic(fmt.Errorf("can't decode public key: %w", err))
+			}
+			app.valAddrToPubKeyMap[string(pubkey.Address())] = v.PubKey
+		}
+	}
+	return &abci.ResponseInfo{
+		Version:         version.ABCIVersion,
+		AppVersion:      ApplicationVersion,
+		LastBlockHeight: app.state.Height,
+
+		LastBlockAppHash: app.state.Hash(),
+	}, nil
+}
+
+// Query blockchain
+func (app ForumApp) Query(ctx context.Context, query *abci.RequestQuery) (*abci.ResponseQuery, error) {
+	resp := abci.ResponseQuery{Key: query.Data}
+	// Just testing whether validators are properly stored
+	if query.Path == "/val" {
+		for k := range app.valAddrToPubKeyMap {
+			return &types.ResponseQuery{
+				Key:   query.Data,
+				Value: []byte(string(k)),
+			}, nil
+		}
```

Comment:
Won't you return the `ResponseQuery` at the first iteration of the loop, no matter how many validators there are?

Reply:
I guess there's an "if" missing...

Reply:
Yes, you are right, this was a quick test which I forgot to update. This should be actually read from the DB.